### PR TITLE
Improve Developer Testing Experience

### PR DESCRIPTION
**Pull request by omad**
# Background
See #521 

# Changes in this PR
- Add [pytest fixtures](https://docs.pytest.org/en/6.2.x/fixture.html) which automatically start and stop a PostgreSQL Docker Container, and provision an ODC Database containing Metadata Types and Products. This is just for the `odc.apps` tests.
- Use [moto](http://docs.getmoto.org/en/latest/) to mock AWS services instead of relying on external S3 objects, which have recently changed and broken tests.
- Remove [pytest-depends](https://pypi.org/project/pytest-depends/) and replace with [pytest fixtures](https://docs.pytest.org/en/6.2.x/fixture.html) , making it possible to run any test individually.
- Unfortunately: Lots of linting and formatting related fixes

### CODE REVIEW

**Comment by omad** [comment]
As described in #526 I'm going to fix the outstanding linting issues caught by pre-commit.ci in a follow up PR.

File: apps/dc_tools/odc/apps/dc_tools/utils.py [hunk 1]
```diff
@@ -273,7 +269,7 @@ def index_update_dataset(
                     updates = {tuple(): changes.allow_any}
                 # Do the updating
                 try:
-                    dc.index.datasets.update(ds, updates_allowed=updates)
+                    dc.index.datasets.update_flag(ds, updates_allowed=updates)
```
Comment (Ariana-B) [inline apps/dc_tools/odc/apps/dc_tools/utils.py]: accidental replace?

File: apps/dc_tools/tests/conftest.py [hunk 2]
```diff
@@ -81,12 +169,144 @@ def final_dsid():
...
+def odc_test_db_with_products(odc_db: Datacube):
+
+    local_csv = str(Path(__file__).parent / "data/example_product_list.csv")
+    added, updated, failed = add_update_products(odc_db, local_csv)
+
+    assert failed == 0
+    yield odc_db
 
 
 @pytest.fixture
```
Comment (Ariana-B) [inline apps/dc_tools/tests/conftest.py:308]: Should this fixture also be updated?

File: libs/cloud/odc/aws/_find.py [hunk 3]
```diff
@@ -64,7 +66,7 @@ def is_glob(s):
         elif "**" not in qq_set:
             depth = len(qq)
         else:
-            raise ValueError("Bad query: %s" % url_query)
+            raise ValueError("Bad query: {url_query}")
```
Comment (Ariana-B) [inline libs/cloud/odc/aws/_find.py]: Missing an `f` in front

**Review by Ariana-B — DISMISSED** [review]
Let's either determine what's causing the pylint warning with `update` and `update-if-exists` and revert the name change, or update the `cli` parameters in each file accordingly

**Review by Ariana-B — COMMENTED** [review]
Other than the couple minor comments, looks good!

File: apps/dc_tools/tests/conftest.py (hunk 2 above)
Comment (omad) [reply apps/dc_tools/tests/conftest.py:308]: Yep, thanks!

**Comment by omad** [comment]
Thanks for catching those Ariana!

**Pull request merged**